### PR TITLE
search/cve: exclude unsupported images from fixed-tag list.

### PR DESCRIPTION
**Pull request by shimish2**
Previously only image validation was done for only first tag and assumed that all tags will have same image media type but this was wrong assumption, changed the logic to enable image validation specific to image tag.
Also added fix for issue #130, if image scanning not supported due to invalid media type adding that image as an infected image.

### CODE REVIEW

File: pkg/extensions/search/cve/cve.go [hunk 1]
```diff
@@ -121,79 +97,55 @@ func dirExists(d string) bool {
...
-		if err != nil {
-			cveinfo.Log.Error().Err(err).Msg("Unable to open Image Layers file")
-
-			return nil, err
-		}
-
-		if err := json.Unmarshal(blobBuf, &layerIndex); err != nil {
-			cveinfo.Log.Error().Err(err).Msg("Unable to marshal blob index")
+			if err != nil {
+				return tagsInfo, err
```
Comment (rchincha) [inline pkg/extensions/search/cve/cve.go:139]: Do you want to log here?

File: pkg/extensions/search/cve/cve.go [hunk 2]
```diff
@@ -45,53 +45,29 @@ func ScanImage(config *config.Config) (report.Results, error) {
...
-	var index ispec.Index
-
-	var blobManifest v1.Manifest
-
-	var digest godigest.Digest
-
-	if err := json.Unmarshal(buf, &index); err != nil {
-		cveinfo.Log.Error().Err(err).Msg("Unable to marshal index.json file")
-
 		return false, err
```
Comment (rchincha) [inline pkg/extensions/search/cve/cve.go:59]: log here?

File: pkg/extensions/search/cve/cve.go [hunk 3]
```diff
@@ -224,3 +182,66 @@ func GetFixedTags(allTags []TagInfo, infectedTags []TagInfo) []TagInfo {
 
 	return fixedTags
 }
+
+func (cveinfo CveInfo) getImageManifests(imagePath string) ([]ispec.Descriptor, error) {
+	buf, err := ioutil.ReadFile(path.Join(imagePath, "index.json"))
+
+	if err != nil {
+		if os.IsNotExist(err) {
+			cveinfo.Log.Error().Err(err).Msg("Index.json does not exist")
```
Comment (rchincha) [inline pkg/extensions/search/cve/cve.go:191]: index.json , lowercase

File: pkg/extensions/search/resolver.go [hunk 4]
```diff
@@ -188,6 +162,13 @@ func (r *queryResolver) ImageListForCve(ctx context.Context, id string) ([]*ImgR
 		for _, tag := range tagList {
 			r.cveInfo.CveTrivyConfig.TrivyConfig.Input = path.Join(r.dir, repo+":"+tag)
 
+			isValidImage, _ := r.cveInfo.IsValidImageFormat(r.cveInfo.CveTrivyConfig.TrivyConfig.Input)
+			if !isValidImage {
+				r.cveInfo.Log.Debug().Str("Image media type not supported for scanning", repo)
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:167]: .Msg() missing? else it will not log

File: pkg/extensions/search/cve/cve.go (hunk 3 above)
Comment (shimish2) [reply pkg/extensions/search/cve/cve.go:191]: Ok, i will make it to lowercase

File: pkg/extensions/search/resolver.go (hunk 4 above)
Comment (shimish2) [reply pkg/extensions/search/resolver.go:167]: Yes, it will not log. I will change it.

File: pkg/extensions/search/cve/cve.go [hunk 5]
```diff
@@ -45,53 +45,29 @@ func ScanImage(config *config.Config) (report.Results, error) {
 }
 
 func (cveinfo CveInfo) IsValidImageFormat(imagePath string) (bool, error) {
-	imageDir := getImageDir(imagePath)
+	imageDir, inputTag := getImageDirAndTag(imagePath)
 
 	if !dirExists(imageDir) {
-		cveinfo.Log.Error().Msg("Image Directory not exists")
+		cveinfo.Log.Error().Msg("image directory not exists")
```
Comment (rchincha) [inline pkg/extensions/search/cve/cve.go:51]: "not exists" -> "doesn't exist"

File: pkg/extensions/search/cve/cve.go [hunk 6]
```diff
@@ -224,3 +182,66 @@ func GetFixedTags(allTags []TagInfo, infectedTags []TagInfo) []TagInfo {
 
 	return fixedTags
 }
+
+func (cveinfo CveInfo) getImageManifests(imagePath string) ([]ispec.Descriptor, error) {
+	buf, err := ioutil.ReadFile(path.Join(imagePath, "index.json"))
+
+	if err != nil {
+		if os.IsNotExist(err) {
+			cveinfo.Log.Error().Err(err).Msg("index.json does not exist")
```
Comment (rchincha) [inline pkg/extensions/search/cve/cve.go:191]: "doesn't" or "is missing" but be consistent everywhere

File: pkg/extensions/search/resolver.go [hunk 7]
```diff
@@ -53,24 +52,18 @@ func GetResolverConfig(dir string, log log.Logger, imgstorage *storage.ImageStor
...
-		r.cveInfo.Log.Error().Err(err).Msg("Error scanning image repository")
+		r.cveInfo.Log.Debug().Msg("image media type not supported for scanning")
 
 		return &CVEResultForImage{}, err
 	}
 
 	cveResults, err := cveinfo.ScanImage(r.cveInfo.CveTrivyConfig)
 	if err != nil {
-		r.cveInfo.Log.Error().Err(err).Msg("Error scanning image repository")
+		r.cveInfo.Log.Error().Err(err).Msg("error scanning image repository")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:66]: "unable to scan ..."

File: pkg/extensions/search/cve/cve.go (hunk 5 above)
Comment (shimish2) [reply pkg/extensions/search/cve/cve.go:51]: Ok

File: pkg/extensions/search/resolver.go [hunk 8]
```diff
@@ -143,42 +136,23 @@ func (r *queryResolver) CVEListForImage(ctx context.Context, image string) (*CVE
 func (r *queryResolver) ImageListForCve(ctx context.Context, id string) ([]*ImgResultForCve, error) {
 	cveResult := []*ImgResultForCve{}
 
-	r.cveInfo.Log.Info().Msg("Extracting Repositories")
+	r.cveInfo.Log.Info().Msg("extracting repositories")
 
 	repoList, err := r.imgStore.GetRepositories()
 	if err != nil {
-		r.cveInfo.Log.Error().Err(err).Msg("Not able to search repositories")
+		r.cveInfo.Log.Error().Err(err).Msg("not able to search repositories")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:143]: "unable to ..."

File: pkg/extensions/search/resolver.go [hunk 9]
```diff
@@ -143,42 +136,23 @@ func (r *queryResolver) CVEListForImage(ctx context.Context, image string) (*CVE
...
-			r.cveInfo.Log.Error().Err(err).Str("Error reading image media type", repo)
-
-			return cveResult, err
-		}
+		r.cveInfo.Log.Info().Str("extracting list of tags available in image", repo).Msg("")
 
 		tagList, err := r.imgStore.GetImageTags(repo)
 		if err != nil {
-			r.cveInfo.Log.Error().Err(err).Msg("Not able to get list of Image Tag")
+			r.cveInfo.Log.Error().Err(err).Msg("not able to get list of image tag")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:155]: "unable to ..."

File: pkg/extensions/search/resolver.go [hunk 10]
```diff
@@ -188,11 +162,18 @@ func (r *queryResolver) ImageListForCve(ctx context.Context, id string) ([]*ImgR
 		for _, tag := range tagList {
 			r.cveInfo.CveTrivyConfig.TrivyConfig.Input = path.Join(r.dir, repo+":"+tag)
 
-			r.cveInfo.Log.Info().Str("Scanning Image", path.Join(r.dir, repo+":"+tag)).Msg("")
+			isValidImage, _ := r.cveInfo.IsValidImageFormat(r.cveInfo.CveTrivyConfig.TrivyConfig.Input)
+			if !isValidImage {
+				r.cveInfo.Log.Debug().Str("image media type not supported for scanning", repo+":"+tag).Msg("")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:167]: be consistent with the error msg elsewhere

File: pkg/extensions/search/resolver.go [hunk 11]
```diff
@@ -188,11 +162,18 @@ func (r *queryResolver) ImageListForCve(ctx context.Context, id string) ([]*ImgR
...
+
+				continue
+			}
+
+			r.cveInfo.Log.Info().Str("scanning image", path.Join(r.dir, repo+":"+tag)).Msg("")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:172]: Info().Str("path", path.Join(...)).Msg("scanning image")

File: pkg/extensions/search/resolver.go [hunk 12]
```diff
@@ -188,11 +162,18 @@ func (r *queryResolver) ImageListForCve(ctx context.Context, id string) ([]*ImgR
...
 
 			results, err := cveinfo.ScanImage(r.cveInfo.CveTrivyConfig)
 			if err != nil {
-				r.cveInfo.Log.Error().Err(err).Str("Error scanning image", repo+":"+tag)
+				r.cveInfo.Log.Error().Err(err).Str("error scanning image", repo+":"+tag)
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:176]: "unable to scan image"

File: pkg/extensions/search/resolver.go [hunk 13]
```diff
@@ -222,22 +203,11 @@ func (r *queryResolver) ImageListForCve(ctx context.Context, id string) ([]*ImgR
...
-	if !isValidImage {
-		r.cveInfo.Log.Debug().Msg("Image media type not supported for scanning")
-
-		return imgResultForFixedCVE, errors.ErrScanNotSupported
-	}
-
-	if err != nil {
-		return imgResultForFixedCVE, err
-	}
+	r.cveInfo.Log.Info().Str("extracting list of tags available in image", image).Msg("")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:206]: Put that in the Msg() part. Str() etc are for args.

File: pkg/extensions/search/resolver.go [hunk 14]
```diff
@@ -222,22 +203,11 @@ func (r *queryResolver) ImageListForCve(ctx context.Context, id string) ([]*ImgR
...
 
 	tagsInfo, err := r.cveInfo.GetImageTagsWithTimestamp(r.dir, image)
 	if err != nil {
-		r.cveInfo.Log.Error().Err(err).Msg("Error while readling image media type")
+		r.cveInfo.Log.Error().Err(err).Msg("error while readling image tags")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:210]: "unable to read image tags"

File: pkg/extensions/search/resolver.go [hunk 15]
```diff
@@ -249,11 +219,20 @@ func (r *queryResolver) ImageListWithCVEFixed(ctx context.Context, id string, im
 	for _, tag := range tagsInfo {
 		r.cveInfo.CveTrivyConfig.TrivyConfig.Input = path.Join(r.dir, image+":"+tag.Name)
 
-		r.cveInfo.Log.Info().Str("Scanning image", path.Join(r.dir, image+":"+tag.Name)).Msg("")
+		isValidImage, _ := r.cveInfo.IsValidImageFormat(r.cveInfo.CveTrivyConfig.TrivyConfig.Input)
+		if !isValidImage {
+			r.cveInfo.Log.Debug().Msg("image media type not supported for scanning, adding as a infected image")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:224]: "adding as an affected image"

File: pkg/extensions/search/resolver.go [hunk 16]
```diff
@@ -249,11 +219,20 @@ func (r *queryResolver) ImageListWithCVEFixed(ctx context.Context, id string, im
...
+
+			continue
+		}
+
+		r.cveInfo.Log.Info().Str("scanning image", path.Join(r.dir, image+":"+tag.Name)).Msg("")
 
 		results, err := cveinfo.ScanImage(r.cveInfo.CveTrivyConfig)
 		if err != nil {
-			r.cveInfo.Log.Error().Err(err).Str("Error scanning image", image+":"+tag.Name).Msg("")
+			r.cveInfo.Log.Error().Err(err).Str("error scanning image", image+":"+tag.Name).Msg("")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:235]: Str("image", image).Msg("unable to scan image")

File: pkg/extensions/search/resolver.go [hunk 17]
```diff
@@ -275,27 +254,35 @@ func (r *queryResolver) ImageListWithCVEFixed(ctx context.Context, id string, im
...
 	if len(infectedTags) != 0 {
-		r.cveInfo.Log.Info().Msg("Comparing fixed tags timestamp")
+		r.cveInfo.Log.Info().Msg("comparing fixed tags timestamp")
 
 		fixedTags := cveinfo.GetFixedTags(tagsInfo, infectedTags)
 
-		finalTagList := make([]*TagInfo, 0)
+		finalTagList = getGraphqlCompatibleTags(fixedTags)
+	} else {
+		r.cveInfo.Log.Info().Msg("input image does not contain any tag that have given cve")
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:266]: Which image? add that as a Str() arg

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**